src/main/scala/beam/agentsim/agents/ridehail/MatchmakingUtils.scala

```diff
@@ -0,0 +1,394 @@
+      val mainRequests = customers.slice(0, Math.min(customers.size, v.getSeatingCapacity))
+      mainRequests ::: customers
+        .drop(mainRequests.size)
+        .filter(
+          r =>
+            mainRequests.foldLeft(false) { (acc, t) =>
+              acc || MatchmakingUtils
+                .checkAngle(t.pickup.activity.getCoord, t.dropoff.activity.getCoord, r.dropoff.activity.getCoord)
+          }
+        )
+    }
+  }
+
+  def getRidehailSchedule(
+    schedule: List[MobilityRequest],
+    newRequests: List[MobilityRequest],
+    remainingVehicleRangeInMeters: Int,
+    skimmer: BeamSkimmer
+  ): Option[List[MobilityRequest]] = {
+    val newPoolingList = scala.collection.mutable.ListBuffer.empty[MobilityRequest]
+    val reversedSchedule = schedule.reverse
+    val sortedRequests = reversedSchedule.lastOption match {
+      case Some(_) if reversedSchedule.exists(_.tag == EnRoute) =>
+        val enRouteIndex = reversedSchedule.indexWhere(_.tag == EnRoute) + 1
+        newPoolingList.appendAll(reversedSchedule.slice(0, enRouteIndex))
+        // We make sure that request time is always equal or greater than the driver's "current tick" as denoted by time in EnRoute
+        val shiftRequestsBy =
+          Math.max(0, reversedSchedule(enRouteIndex - 1).baselineNonPooledTime - newRequests.head.baselineNonPooledTime)
+        (reversedSchedule.slice(enRouteIndex, reversedSchedule.size) ++ newRequests.map(
+          req =>
+            req.copy(
+              baselineNonPooledTime = req.baselineNonPooledTime + shiftRequestsBy,
+              serviceTime = req.serviceTime + shiftRequestsBy
+          )
+        )).sortBy(
+          mr => (mr.baselineNonPooledTime, mr.person.map(_.personId.toString).getOrElse("ZZZZZZZZZZZZZZZZZZZZZZZ"))
+        )
+      case Some(_) =>
+        newPoolingList.appendAll(reversedSchedule)
+        newRequests.sortBy(_.baselineNonPooledTime)
+      case None =>
+        val temp = newRequests.sortBy(_.baselineNonPooledTime)
+        newPoolingList.append(temp.head)
+        temp.drop(1)
+    }
+    sortedRequests.foreach { curReq =>
+      val prevReq = newPoolingList.lastOption.getOrElse(newPoolingList.last)
+      val tdc = skimmer.getTimeDistanceAndCost(
+        prevReq.activity.getCoord,
+        curReq.activity.getCoord,
+        prevReq.baselineNonPooledTime,
+        BeamMode.CAR,
+        Id.create(
+          skimmer.beamScenario.beamConfig.beam.agentsim.agents.rideHail.initialization.procedural.vehicleTypeId,
+          classOf[BeamVehicleType]
+        )
+      )
+      val serviceTime = prevReq.serviceTime + tdc.time
+      val serviceDistance = prevReq.serviceDistance + tdc.distance.toInt
+      if (serviceTime <= curReq.upperBoundTime && serviceDistance <= remainingVehicleRangeInMeters) {
+        newPoolingList.append(curReq.copy(serviceTime = serviceTime, serviceDistance = serviceDistance))
+      } else {
+        return None
```

Comment:
Another return. Please find another way to solve this. I would go with `forall` instead:

```
val isValid = sortedRequests.forall(curReq =>
    ...
    if (serviceTime <= curReq.upperBoundTime && serviceDistance <= remainingVehicleRangeInMeters) {
        newPoolingList.append(curReq.copy(serviceTime = serviceTime, serviceDistance = serviceDistance))
        true
    } else {
        false
    }
)
if isValid Some(newPoolingList.toList) else None
```